### ISSUE TITLE
Add feature to support 'Audit' style logging

### ISSUE THREAD

**Issue by heavytread** [serilog/serilog#820]
A feature should be added to support scenarios in which the calling code needs to confirm that an event has been logged to a sink and committed to the underlying store before proceeding with some action.

For example, say we are building an application for a government agency that handles records containing personally identifiable information (PII). Whenever data is updated, or whenever a certain query type is performed (e.g.: for a record containing personal info), the system will need to write an audit log that records the ID of the user who took the action. In such a case, the system needs to know that the audit record was written before allowing the user (or, in the case of a service, the caller) to proceed with the requested action. If, for whatever reason, the audit record cannot be written, the requested action must fail. I.e.: the audit record is *required* as part of the requested operation.

There are currently two features of Serilog that together make scenarios like this difficult (or impossible) to implement:

1) Sinks can optionally derive from ```PeriodicBatchingSink```. This sink allows events to be deferred and written to the underlying data store in bulk. Because commits to the underlying data store can be deferred, there is no way for applications to ensure that audit events have been written before proceeding with an audited operation

2) Serilog eats all exceptions thrown by sinks. The ```SafeAggregateSink``` and ```SafeAggregateEncricher``` classes wrap calls to sinks and enrichers in a catch block that protects applications from exceptions thrown during logging. This prevents apps from knowing when an attempt to write an audit log record has failed.

A simple workaround would be to provide developers with the option to defeat the behaviors described above, for instance by providing a method on ```LoggerConfiguration``` that can be called before the call to ```CreateLogger()```.

**Comment by jods4** [comment]
:+1:
I was looking into replacing NLog with Serilog in our projects and this was one shortcoming of Serilog.
We have a kind of "audit" log beside our normal application log and the exact requirements described here:
- Non-buffered writes;
- Throwing when logging fails.

NLog has options to support both of these.

**Comment by nblumhardt** [comment]
Thanks for the input @heavytread and @jods4 - good to know this is a desirable feature for some folks. Some early exploration was done in this space but there wasn't a lot of demand so it was put on hold.

There are a few design possibilities with pros and cons; with more time to think about it, I'm back to preferring:

```csharp
var logger = new LoggerConfiguration()
    .AuditTo.File("audit.txt")
    .WriteTo.SomeOtherSink()
    .CreateLogger();
```

In this model, the `LoggerConfiguration` structures the internal pipeline so that `Safe*` is not applied to  any `Audit` sinks, but does still wrap regular `WriteTo` sinks so that events from the audit log can be (unreliably) mirrored to something like Seq for easy inspection.

While `AuditTo` would ultimately boil down to adding a regular `ILogEventSink`, the configuration object would be `LoggerAuditConfiguration` instead of `LoggerSinkConfiguration`, forcing sinks that intrinsically supported auditing to opt in by providing an additional extension method on this type. This would permit sinks like those based on `PeriodicBatchingSink` to configure an alternative internal implementation for auditing scenarios, as the batch/timer-based implementation is unlikely to provide audit-like guarantees.

In practice I think this'd all work fairly smoothly because nearly all auditing will rely on `File`, `RollingFile`, `EventLog` or one of a few other reliable-enough sinks.

In the situation where a single logger might be used across the whole app, but only a subset of events "audited", I think sub-loggers would work here:

```csharp
var logger = new LoggerConfiguration()
    .AuditTo.Logger(lc => lc
        .Filter.ByIncludingOnly(Matching.WithProperty("IsAudited", true))
        .AuditTo.File("audit.txt"))
    .WriteTo.SomeOtherSink()
    .CreateLogger();
```

How does this proposal sound?

**Comment by jods4** [comment]
sounds good!

> In practice I think this'd all work fairly smoothly because nearly all auditing will rely on  File ,  RollingFile , 

Yes but you'd be amazed what this kind of config helped us caught with NLog. 
Sometimes there's a typo in the configured path, the ACL on the file or folder are incorrect, the disk is full, the file is locked by another process or the network share is not available...

Everything usually works smoothly but it's terrible when the day you need to look at a log (or worse an audit log) you notice you had a bad config and there's no log!

**Comment by nblumhardt** [comment]
Thanks @jods4 I'll run with this then :+1:

**Comment by heavytread** [comment]
Hi Nicholas. Yes, all of that sounds very clean and a lot less hacky than what I'd proposed. Forcing sinks to declare whether or not they support auditing sounds lot better than changing the batching behavior so that batching might not actually happen even when the sink developer is expecting it to.

**Comment by nblumhardt** [comment]
@heavytread this is now available in the published `-dev` packages for Serilog, Serilog.Sinks.File and Serilog.Sinks.Seq. We've got a tracking issue #828 to cover Serilog.Sinks.RollingFile and the two configuration subsystems.

Is it likely you'll have a chance to try out what's currently in the `-dev` build? Some additional validation/feedback would be great :-)

**Comment by normanhh3** [comment]
Call me crazy, but that sounds like it really is starting to cross down the path of event sourcing and making Serilog an [event sourcing implementation](https://github.com/EventStore/EventStore) seems to overload its value as a logging platform.  That said, it would be interesting to see how the event sourcing practice and logging would best complement one another.

**Comment by nblumhardt** [comment]
@normanhh3 you could say that :-) but event sourcing is definitely not the intent here. Another way of thinking about it is that audit events are almost like a super-critical level, `SoImportantThatINeedToMakeSureThisReachesDiskBeforeMovingOn`. Many ways to use it, I suppose.

Anyhow, this one shipped in 2.2 (IIRC) so the ticket here is out-of-date. Cheers!
